### PR TITLE
[INFRA] Move core/algorithm into core/configuration

### PR DESCRIPTION
**Pull request by Irallia**
Resolves partly https://github.com/seqan/product_backlog/issues/160

algorithm/concept.hpp -> core/configuration/detail/.
algorithm/configuration.hpp -> core/configuration/.
algorithm/configuration_element_debug_mode.hpp -> core/configuration/detail/.
algorithm/configuration_element_parallel_mode.hpp -> core/configuration/detail/.
algorithm/configuration_utility.hpp -> core/configuration/detail/.

Skipped: algorithm/pipeable_config_element.hpp (needs a decision)

Keep untouched:
algorithm/detail/... -> core/algorithm/detail/...
algorithm/algorithm_result_generator_range -> core/algorithm/algorithm_result_generator_range

Added new all.hpp's and adjusted the other ones.

Details for this: https://gist.github.com/rrahn/f6b09ad67bebd14bd9c58f77c53e0450

### CODE REVIEW

File: include/seqan3/core/algorithm/all.hpp [hunk 1]
```diff
@@ -15,67 +15,12 @@
...
 #include <seqan3/core/algorithm/algorithm_result_generator_range.hpp>
-#include <seqan3/core/algorithm/bound.hpp>
-#include <seqan3/core/algorithm/concept.hpp>
-#include <seqan3/core/algorithm/configuration_utility.hpp>
-#include <seqan3/core/algorithm/configuration.hpp>
-#include <seqan3/core/detail/pack_algorithm.hpp>
+#include <seqan3/core/algorithm/detail/all.hpp>
 #include <seqan3/core/algorithm/pipeable_config_element.hpp>
+#include <seqan3/core/configuration/configuration.hpp> // ToDo: remove this with the release SeqAn-3.1.0.
+#include <seqan3/core/configuration/configuration_element_debug_mode.hpp> // ToDo: remove this with the release SeqAn-3.1.0.
```
Comment (joergi-w) [inline include/seqan3/core/algorithm/all.hpp]: Do you mean the following?
```suggestion
#include <seqan3/core/configuration/detail/configuration_element_debug_mode.hpp> // ToDo: remove this with the release SeqAn-3.1.0.
```

File: include/seqan3/core/configuration/detail/all.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Meta-header for the detail functionality of core configuration detail.
```
Comment (joergi-w) [inline include/seqan3/core/configuration/detail/all.hpp:9]: This is somehow double-expressed and I suggest:
```suggestion
 * \brief Meta-header for the detail functionality of configuration.
```

File: include/seqan3/core/configuration/detail/all.hpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
```
Comment (joergi-w) [inline include/seqan3/core/configuration/detail/all.hpp:8]: Isn't it a problem that we have all this nice documentation here but as it is in a detail header it is invisible to users?

**Review by joergi-w — CHANGES_REQUESTED** [review]

File: include/seqan3/core/configuration/detail/concept.hpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+#pragma once
+
+#include <functional>
+#include <type_traits>
+
+#include <meta/meta.hpp>
+
+#include <seqan3/core/type_traits/template_inspection.hpp>
+#include <seqan3/std/concepts>
```
Comment (joergi-w) [inline include/seqan3/core/configuration/detail/concept.hpp:21]: Please move this header up to the std-includes and sort it as if its name is `<concepts>`

File: include/seqan3/core/configuration/detail/configuration_utility.hpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2020, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2020, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides functionality to access get function by enum values.
```
Comment (joergi-w) [inline include/seqan3/core/configuration/detail/configuration_utility.hpp:9]: This description seems to be wrong here... at least I don't understand the relation to the actual code in this file.

**Review by joergi-w — CHANGES_REQUESTED** [review]

File: include/seqan3/core/configuration/detail/all.hpp (hunk 3 above)
Comment (Irallia) [reply include/seqan3/core/configuration/detail/all.hpp:8]: Yes, you're right. I move this to the configutation/all.hpp

File: include/seqan3/core/algorithm/concept.hpp [hunk 6]
```diff
@@ -6,111 +6,15 @@
...
-#else // ^^^ workaround / no workaround vvv
-    { config_t::id };
-#endif // SEQAN3_WORKAROUND_GCC_PIPEABLE_CONFIG_CONCEPT
-};
-//!\endcond
-
-} // namespace seqan3::detail
+SEQAN3_DEPRECATED_HEADER(
+   "This header is deprecated and will be removed in SeqAn-3.1.0; Please #include "
+   "<seqan3/core/configuration/detail/concept.hpp> instead.")
```
[the remark this answers is not in the document]
Comment (Irallia) [reply include/seqan3/core/algorithm/concept.hpp:20]: Hmm, I don't know. I would say this message is usefull for every developer and will anyways get removed with the next big release. But lets ask someone of the @seqan/core team.

[the remark this answers is not in the document]
Comment (eseiler) [reply include/seqan3/core/algorithm/concept.hpp:20]: The idea is that if you use old includes, you get deprecation warnings, but nothing breaks.
So you need to include the new headers in the old header and add a deprecation note in the old header.

We do not differentiate between detail and non-detail headers for this.

[the remark this answers is not in the document]
Comment (joergi-w) [reply include/seqan3/core/algorithm/concept.hpp:20]: Okay 👍

**Review by joergi-w — APPROVED** [review]
Thank you!

**Review by rrahn — CHANGES_REQUESTED** [review]
Please do not remove the old seqan3/alignment/band in this PR. This has nothing to do with the change of the configuration header. Second the static band should be deprecated before it is removed in 3.1. Can you first revert these changes and then I look into the rest. Thank you

File: include/seqan3/core/algorithm/all.hpp [hunk 7]
```diff
@@ -15,67 +15,9 @@
...
- *
- * \include test/snippet/core/algorithm/configuration_get_or.cpp
  */
 #include <seqan3/core/algorithm/algorithm_result_generator_range.hpp>
 #include <seqan3/core/algorithm/bound.hpp>
-#include <seqan3/core/algorithm/concept.hpp>
-#include <seqan3/core/algorithm/configuration_utility.hpp>
-#include <seqan3/core/algorithm/configuration.hpp>
-#include <seqan3/core/detail/pack_algorithm.hpp>
+#include <seqan3/core/algorithm/detail/all.hpp>
```
Comment (rrahn) [inline include/seqan3/core/algorithm/all.hpp:21]: please remove

File: include/seqan3/core/algorithm/all.hpp [hunk 8]
```diff
@@ -15,67 +15,9 @@
...
 #include <seqan3/core/algorithm/pipeable_config_element.hpp>
+#include <seqan3/core/configuration/all.hpp> // ToDo: remove this with the release SeqAn-3.1.0.
```
Comment (rrahn) [inline include/seqan3/core/algorithm/all.hpp:23]: Not sure how to handle this. This is likely to be forgotten. 
I am thinking of this:
```cpp
#if SEQAN3_VERSION_MAJOR == 3 && SEQAN3_VERSION_MINOR == 1
  #pragma warning "Remove #include <seqan3/core/configuration/all.hpp> from this header."
#endif
```
@marehr what do you think?

File: include/seqan3/core/algorithm/concept.hpp [hunk 9]
```diff
@@ -6,111 +6,15 @@
 // -----------------------------------------------------------------------------------------------------
 
 /*!\file
- * \brief Provides concepts for the configuration classes.
+ * \brief [DEPRECATED] Provides concepts for the configuration classes.
  * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ * \deprecated This header will be removed in 3.1. Please use seqan3/core/configuration/detail/concept.hpp instead.
```
Comment (rrahn) [inline include/seqan3/core/algorithm/concept.hpp:11]:
```suggestion
 * \deprecated This header will be removed in 3.1.0; Please \#include <seqan3/core/configuration/detail/concept.hpp> instead.
```

File: include/seqan3/core/algorithm/configuration_element_debug_mode.hpp [hunk 10]
```diff
@@ -6,41 +6,16 @@
 // -----------------------------------------------------------------------------------------------------
 
 /*!\file
- * \brief Provides seqan3::detail::debug_mode.
+ * \brief [DEPRECATED] Provides seqan3::detail::debug_mode.
  * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ * \deprecated This header will be removed in 3.1. Please use
+ *             seqan3/core/configuration/detail/configuration_element_debug_mode.hpp instead.
```
Comment (rrahn) [inline include/seqan3/core/algorithm/configuration_element_debug_mode.hpp:12]:
```suggestion
 * \deprecated This header will be removed in 3.1.0; Please \#include <seqan3/core/configuration/detail/configuration_element_debug_mode.hpp> instead.
```

File: include/seqan3/core/algorithm/configuration_element_debug_mode.hpp [hunk 11]
```diff
@@ -6,41 +6,16 @@
...
-    ~debug_mode() = default; //!< Defaulted.
-    //!\}
-
-    //!\brief Internal id to check for consistent configuration settings.
-    static constexpr typename wrapped_config_id_t::value_type id{wrapped_config_id_t::value};
-};
-} // namespace seqan3::detail
+SEQAN3_DEPRECATED_HEADER(
+   "This header is deprecated and will be removed in SeqAn-3.1.0; Please #include "
+   "<seqan3/core/configuration/detail/configuration_element_debug_mode.hpp> instead.")
```
Comment (rrahn) [inline include/seqan3/core/algorithm/configuration_element_debug_mode.hpp:21]:
```suggestion
   "This header is deprecated and will be removed in SeqAn-3.1.0; Please #include <seqan3/core/configuration/detail/configuration_element_debug_mode.hpp> instead.")
```
We said that it is ok to not wrap this around. I think it causes some displaying issues.

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp [hunk 12]
```diff
@@ -6,53 +6,17 @@
 // -----------------------------------------------------------------------------------------------------
 
 /*!\file
- * \brief Provides seqan3::detail::parallel_mode.
+ * \brief [DEPRECATED] Provides seqan3::detail::parallel_mode.
  * \author Svenja Mehringer <svenja.mehringer AT fu-berlin.de>
  * \author Jörg Winkler <j.winkler AT fu-berlin.de>
+ * \deprecated This header will be removed in 3.1. Please use
+ *             seqan3/core/configuration/detail/configuration_element_parallel_mode.hpp instead.
```
Comment (rrahn) [inline include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp:13]:
```suggestion
 * \deprecated This header will be removed in 3.1.0; Please \#include
 *             <seqan3/core/configuration/detail/configuration_element_parallel_mode.hpp> instead.
```

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp [hunk 13]
```diff
@@ -6,53 +6,17 @@
...
-
-    /*!\privatesection
-     * \brief Internal id to check for consistent configuration settings.
-     */
-    static constexpr typename wrapped_config_id_t::value_type id{wrapped_config_id_t::value};
-};
-} // namespace seqan3::detail
+SEQAN3_DEPRECATED_HEADER(
+   "This header is deprecated and will be removed in SeqAn-3.1.0; Please #include "
+   "<seqan3/core/configuration/detail/configuration_element_parallel_mode.hpp> instead.")
```
Comment (rrahn) [inline include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp:22]:
```suggestion
   "This header is deprecated and will be removed in SeqAn-3.1.0; Please #include <seqan3/core/configuration/detail/configuration_element_parallel_mode.hpp> instead.")
```

File: include/seqan3/core/algorithm/configuration_utility.hpp [hunk 14]
```diff
@@ -6,78 +6,16 @@
 // -----------------------------------------------------------------------------------------------------
 
 /*!\file
- * \brief Provides functionality to access get function by enum values.
+ * \brief [DEPRECATED] Provides various auxiliary functions with which parts of the configurations can be checked.
  * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ * \deprecated This header will be removed in 3.1. Please use seqan3/core/configuration/detail/configuration_utility.hpp
+ *             instead.
```
Comment (rrahn) [inline include/seqan3/core/algorithm/configuration_utility.hpp:12]:
```suggestion
 * \deprecated This header will be removed in 3.1.0; Please \#include <seqan3/core/configuration/detail/configuration_utility.hpp>
 *             instead.
```

File: include/seqan3/core/algorithm/configuration_utility.hpp [hunk 15]
```diff
@@ -6,78 +6,16 @@
...
-     * \tparam compare_type The type to compare against `query_t`.
-     */
-    template <typename compare_type>
-    using invoke = is_type_specialisation_of<compare_type, query_t>;
-};
-
-} // namespace seqan3::detail
+SEQAN3_DEPRECATED_HEADER(
+   "This header is deprecated and will be removed in SeqAn-3.1.0; Please #include "
+   "<seqan3/core/configuration/detail/configuration_utility.hpp> instead.")
```
Comment (rrahn) [inline include/seqan3/core/algorithm/configuration_utility.hpp:21]:
```suggestion
   "This header is deprecated and will be removed in SeqAn-3.1.0; Please #include <seqan3/core/configuration/detail/configuration_utility.hpp> instead.")
```

File: include/seqan3/core/algorithm/detail/all.hpp [hunk 16]
```diff
@@ -13,4 +13,5 @@
 #pragma once
 
 #include <seqan3/core/algorithm/detail/algorithm_executor_blocking.hpp>
+#include <seqan3/core/algorithm/detail/execution_handler_parallel.hpp>
```
Comment (rrahn) [inline include/seqan3/core/algorithm/detail/all.hpp:16]: I think in general we don't want detail to be included inside of the all.hpp.
It basically is no functionality for the user and should be included through the public headers.

Comment (rrahn) [reply include/seqan3/core/algorithm/detail/all.hpp:16]: So we leave it for now but I would remove it later. Lets discuss this today in the group meeting

**Review by rrahn — CHANGES_REQUESTED** [review]

File: include/seqan3/core/algorithm/all.hpp (hunk 8 above)
Comment (smehringer) [reply include/seqan3/core/algorithm/all.hpp:23]: This was exactly the problem I tried to explain on Gitter on friday..

File: include/seqan3/core/algorithm/concept.hpp (hunk 9 above)
Comment (Irallia) [reply include/seqan3/core/algorithm/concept.hpp:11]: I followed Marcel's example, which he created for us. (math) Now I had to make a lot of changes that did not match his example. Maybe next time you can look over the example as the core team before we take over all the mistakes there.

File: include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp (hunk 13 above)
Comment (Irallia) [reply include/seqan3/core/algorithm/configuration_element_parallel_mode.hpp:22]: Is your requested change here, that you want everything in one line, even if there are more than 120 characters?

**Comment by Irallia** [comment]
You might want to check my changes before I rebase.

File: include/seqan3/core/algorithm/all.hpp [hunk 17]
```diff
@@ -15,67 +15,10 @@
...
  */
 #include <seqan3/core/algorithm/algorithm_result_generator_range.hpp>
 #include <seqan3/core/algorithm/bound.hpp>
-#include <seqan3/core/algorithm/concept.hpp>
-#include <seqan3/core/algorithm/configuration_utility.hpp>
-#include <seqan3/core/algorithm/configuration.hpp>
-#include <seqan3/core/detail/pack_algorithm.hpp>
 #include <seqan3/core/algorithm/pipeable_config_element.hpp>
+#if SEQAN3_VERSION_MAJOR == 3 && SEQAN3_VERSION_MINOR == 1
+  #pragma warning "Remove #include <seqan3/core/configuration/all.hpp> from this header."
```
Comment (rrahn) [inline include/seqan3/core/algorithm/all.hpp:23]: but where is the header?

**Review by rrahn — COMMENTED** [review]
OK one small thing. I couldn't find the header that should be removed later? The warning is only a message emitted by the compiler. The header include should still exist. Or did I miss it? You can rebase right away.

File: include/seqan3/core/algorithm/all.hpp (hunk 17 above)
Comment (Irallia) [reply include/seqan3/core/algorithm/all.hpp:23]: working without the head... 😅

File: include/seqan3/core/configuration/detail/configuration_utility.hpp [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+#pragma once
+
+#include <array>
+#include <type_traits>
+
+#include <meta/meta.hpp>
+
+#include <seqan3/core/configuration/detail/concept.hpp>
+#include <seqan3/core/concept/tuple.hpp>
```
Comment (rrahn) [inline include/seqan3/core/configuration/detail/configuration_utility.hpp:21]: include order

**Review by rrahn — CHANGES_REQUESTED** [review]
some minor include order then it is good to go

**Review by rrahn — APPROVED** [review]
Thank you

**Pull request merged**